File: src/stack.py
# Stack Implementation in Python
# source: https://www.programiz.com/dsa/stack
from typing import Union

# Creating a stack
def create_stack() -> list:
    stack = []
    return stack

# Check if the stack is empty
def check_empty(stack: list) -> bool:
    return len(stack) == 0

# Adding item(s) into the stack
def push(stack: list, item) -> list:
    stack.append(item)
    print(f'{item} is pushed into stack.')
    print(f'Stack after pushing: {stack}')
    return stack

# Remove an element from stack
def pop(stack: list) -> Union[list, str]:
    if check_empty(stack):
        return 'stack is empty'
    
    return stack.pop()

if __name__ == "__main__":
    stack = create_stack()
    push(stack, 1)
    push(stack, 2)
    push(stack, 3)
    push(stack, 4)
    
    for i in range(len(stack)):
        print(f'Popped item: {pop(stack)}')
        print(f'Stack after popping: {stack}')

File: src/src/stack.cpp
// Stack implementation in C++
// Source: https://www.programiz.com/dsa/stack
#include <stdlib.h>
#include <iostream>

#define MAX 10

int size = 0;

// Creating a stack
struct stack {
    int items[MAX];
    int top;
};

typedef struct stack st;

void createEmptyStack(st *s){
    s->top = -1;
}

// Check if the stack is full
int isFull(st *s) {
    if (s->top == MAX - 1)
        return 1;
    else
        return 0;
}

// Check if the stack is empty
int isEmpty(st *s) {
    if (s->top == -1)
        return 1;
    else
        return 0;
}

// Add element(s) into the stack
void push(st *s, int newItem) {
    if (isFull(s))
        std::cout << "Stack is full" << std::endl;
    else {
        s->top++;
        s->items[s->top] = newItem;
        std::cout << "Item pushed: " << newItem << std::endl;
    }

    size++;
}

// Remove element from stack
void pop(st *s) {
    if (isEmpty(s)){
        std::cout << "Stack is empty" << std::endl;
        return;
    }
    
    else {
        std::cout << "Item popped out: " << s->items[s->top] << std::endl;
        s->top--;
    }
    size--;
}

// Peek top value
int peek(st *s) {
    return s->items[s->top];
}

// Print elements of stack
void printStack(st *s) {
    std::cout << "Stack:" << std::endl;
    for (int i = 0; i < size; i++){
        std::cout << s->items[i] << " ";
    }
    std::cout << std::endl;
}

// Main
int main(){
    int ch;
    st *s = (st *) malloc(sizeof(st));

    createEmptyStack(s);

    for (int i = 1; i < MAX; i++) {
        push(s, i);
    }

    std::cout << "Top element: " << peek(s) << std::endl;

    printStack(s);

    pop(s);

    std::cout << "Top element: " << peek(s) << std::endl;

    std::cout << "\nAfter Popping Out\n";
    printStack(s);
}